### PR TITLE
MM-15452 - Add ability to override LHS icon for bot accounts

### PR DESCRIPTION
**Pull request by ali-farooq0**
#### Summary
Add ability to override LHS icon for bot accounts
- Added a util func `iconImageURLForUser` to get user icon image
- Added `react-inlinesvg` package to support showing SVGs inline
- Updated `SidebarChannelButtonOrLinkIcon` component to show user icon if available, otherwise fallback to default bot icon

#### Ticket Link
Fixes: https://mattermost.atlassian.net/browse/MM-15452

#### Related Pull Requests
- Has server changes (https://github.com/mattermost/mattermost-server/pull/11423)

##### Default Icon
Screen Shot 2019-06-27 at 10 15 00 AM

##### Invision Icon
Screen Shot 2019-06-27 at 10 16 45 AM

##### Github Icon
Screen Shot 2019-06-27 at 10 17 10 AM

##### Default Icon
Screen Shot 2019-06-27 at 10 14 41 AM

##### Invision Icon
Screen Shot 2019-06-27 at 10 01 18 AM

##### Github Icon
Screen Shot 2019-06-27 at 9 58 15 AM

#### Updating live (gif)
2019-06-27 at 4 46 PM

### CODE REVIEW

File: utils/utils.jsx [hunk 1]
```diff
@@ -1339,6 +1340,17 @@ export function imageURLForUser(userIdOrObject) {
     return Client4.getUsersRoute() + '/' + userIdOrObject.id + '/image?_=' + (userIdOrObject.last_picture_update || 0);
 }
 
+export function iconImageURLForUser(userIdOrObject) {
+    if (typeof userIdOrObject == 'string') {
```
Comment (lieut-data) [inline utils/utils.jsx]: nit: It's not an issue here, but can we use `===` for consistency?

File: utils/utils.jsx [hunk 2]
```diff
@@ -1339,6 +1340,17 @@ export function imageURLForUser(userIdOrObject) {
...
+        const profile = getUser(store.getState(), userIdOrObject);
```
Comment (lieut-data) [inline utils/utils.jsx]: There's still a lot of code that uses the global store, but this is considered deprecated as it introduces potential re-rendering issues (the connector won't trigger a re-render for any state queries that occur here).

Can we refactor this to accept the `state` object directly? Also, unless this has strong reuse, I might suggest just defining this function in the connector for the one component that uses it. Easy to refactor out later.

File: components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link_icon.jsx [hunk 3]
```diff
@@ -51,9 +68,21 @@ export default class SidebarChannelButtonOrLinkIcon extends React.PureComponent
...
-                );
+                // Use default bot icon
+                icon = (<BotIcon className='icon icon__bot'/>);
+
+                // Attempt to display custom icon if channelIconUrl has changed
+                // or if there was no error when loading custom svg
+                if ((this.props.channelIconUrl && !this.state.svgError) ||
+                    this.props.channelIconUrl !== this.state.channelIconUrl) {
+                    icon = (
+                        <Svg
```
Comment (lieut-data) [inline components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link_icon.jsx]: What's the experience on load? Will the icon be empty until we fetch it? Is it feasible to render the default icon until it's loaded?

Comment (ali-farooq0) [reply components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link_icon.jsx]: Yes, I actually had the default bot showing while loading the image, but the experience was a bit jittery, so it left it as blank for now.

Comment (lieut-data) [reply components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link_icon.jsx]: Can you elaborate on what jittery means in this context? As in, are we flashing between the icons, or flashing no icon at all?

Comment (ali-farooq0) [reply components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link_icon.jsx]: Jittery in the sense that you momentarily see the default icon which is immediately replaced with the custom icon. Happens really fast. As opposed to No icon and then you see the custom icon appear.

File: utils/utils.jsx (hunk 2 above)
Comment (ali-farooq0) [reply utils/utils.jsx]: Ripped getting user through `getUser` out from here since i really didn't to pass in a `botId`, but rather always a `botuser`.

**Comment by lieut-data** [comment]
@ali-farooq0, one other request to wrap this up: I think the demo plugin uses a bot. Can you update to reflect the override?

**Comment by ali-farooq0** [comment]
> @ali-farooq0, one other request to wrap this up: I think the demo plugin uses a bot. Can you update to reflect the override?

Good idea. Will do.

**Pull request merged**